pkg/rootless/specconv/specconv_linux.go

```diff
@@ -136,3 +156,46 @@ func bindMountHostProcfs(spec *specs.Spec) error {
 
 	return nil
 }
+
+// withBindMountHostIPC replaces /dev/shm and /dev/mqueue  mount with rbind.
+// Required for --ipc=host on rootless.
+//
+// Forked from https://github.com/containerd/nerdctl/blob/v1.1.0/cmd/nerdctl/run.go#L836-L860
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+func bindMountHostIPC(s *specs.Spec) error {
+	for i, m := range s.Mounts {
+		if path.Clean(m.Destination) == "/dev/shm" {
+			newM := specs.Mount{
+				Destination: "/dev/shm",
+				Type:        "bind",
+				Source:      "/dev/shm",
+				Options:     []string{"rbind", "nosuid", "noexec", "nodev"},
+			}
+			s.Mounts[i] = newM
+		}
+		if path.Clean(m.Destination) == "/dev/mqueue" {
+			newM := specs.Mount{
+				Destination: "/dev/mqueue",
+				Type:        "bind",
+				Source:      "/dev/mqueue",
+				Options:     []string{"rbind", "nosuid", "noexec", "nodev"},
+			}
+			s.Mounts[i] = newM
+		}
```

Comment:
Looks like we're doing `path.Clean()` multiple times here, so we should `continue` on the first match. As they're identical in every other way (except for the path, perhaps a switch would be better here? (Looks like we also don't need the intermediate `newM` mount;


```suggestion
		switch p := path.Clean(m.Destination); p {
		case "/dev/shm", "/dev/mqueue":
			s.Mounts[i] = specs.Mount{
				Destination: p,
				Type:        "bind",
				Source:      p,
				Options:     []string{"rbind", "nosuid", "noexec", "nodev"},
			}
		}
```

Reply:
One thing I'm wondering; would it be better to only _patch_ the `Options` instead? (not _replace_?) Could there be other relevant options that were set and that we're now ignoring?

Reply:
☝️ Thinking something like this (perhaps pass by reference would work as well);

```go
func bindMountHostIPC(s *specs.Spec) error {
	for i, m := range s.Mounts {
		switch p := path.Clean(m.Destination); p {
		case "/dev/shm", "/dev/mqueue":
			s.Mounts[i].Options = makeRbind(s.Mounts[i].Options)
		}
	}
	return nil
}

func makeRbind(opts []string) []string {
	for i, o := range opts {
		if o == "rbind" {
			// nothing to do
			break
		}
		if o == "bind" {
			opts[i] = "rbind"
		}
	}
	return opts
}
```

Reply:
That doesn't seem to work
```
$ docker --context=rootless run -it --rm --ipc host alpine
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error mounting "mqueue" to rootfs at "/dev/mqueue": mount mqueue:/dev/mqueue (via /proc/self/fd/7), flags: 0xe: operation not permitted: unknown.
```

Reply:
The first suggestion still works, so I applied it

Reply:
> That doesn't seem to work

Interesting. Perhaps I made a mistake, or something else is missing in that case. Looks like a fun thing to look into, to see what differs (what option is missing).

Thanks for trying it!